### PR TITLE
Migration to jda v5, threads, localized slash commands and permissions

### LINKED ISSUE #138: Implement Support for Threads
When https://github.com/DV8FromTheWorld/JDA/pull/1613 is released we need to support them.

The bot should join threads which are created in reputation channel.
The bot should also listen to messages and command in threads.

### LINKED ISSUE #169: Weekly and monthly leaderboards
Add weekly and monthly rolling leaderboards which shows the reputation of ppl in the last 7 or 30 days

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
- Add support for threads
- Add support for localized slash commands via guild commands
- Add permissions for slash command visibility
- Add top command for month and week
- Remove text commands

Close #138 
Close #169

### CODE REVIEW

File: src/main/resources/locale_fr_FR.properties [hunk 1]
```diff
@@ -192,7 +192,7 @@ command.setup.dialog.starting=Démarre l'installation.
 command.setup.dialog.welcome=Bienvenue dans la configuration du robot de réputation.
 command.thankwords.description=Configurer Thankwords
 command.thankwords.error.patternNotFound=Le modèle n'existe pas.
-command.thankwords.sub.add=Ajouter un autre motif. Les motifs sont interprétés comme des Regex. Un motif ne peut pas contenir d'espaces.
+command.thankwords.sub.add=Add a pattern. These are interpreted as regex. No spaces!
```
Comment (Taucher2003) [inline src/main/resources/locale_fr_FR.properties:195]: English Translation in French

**Review by Taucher2003 — COMMENTED** [review]
As this removes all old message commands, wouldn't it make sense to create a dummy command which just tells that message commands have been deprecated/removed and they should use the slash commands instead?

File: src/main/resources/locale_fr_FR.properties (hunk 1 above)
Comment (RainbowDashLabs) [reply src/main/resources/locale_fr_FR.properties:195]: Lel seems like I copied the wrong one xD

**Comment by RainbowDashLabs** [comment]
> As this removes all old message commands, wouldn't it make sense to create a dummy command which just tells that message commands have been deprecated/removed and they should use the slash commands instead?

Kinda, but I actually hope that ppl are already used to use slash commands already. Doing this would require to keep text command implementations all over the place which is a way too large overhead imo.

Alternative would be to react on all messages with the prefix, but this would probably clash with some other bots, since the repbot ignores unknown commands without an error message.

**Comment by dlsf** [comment]
Will there be a message stating that the bot now uses Slash commands? Because if not, people will probably be confused because the commands are no longer working.

**Comment by RainbowDashLabs** [comment]
> 



> Will there be a message stating that the bot now uses slash commands? Because if not, people will probably be confused because the commands are no longer working.

I mean the bot used slash commands nearly from the start :thinking: We could implement a hacky workaround via another message listener which checks if a command is registered with the name in the command hub. still not nice, but probably a temporary viable solution.

File: src/main/resources/locale_ru_RU.properties [hunk 2]
```diff
@@ -242,7 +243,7 @@ listener.reaction.removal=Репутация выданная %DONOR% была 
 message.welcome=Спасибо за приглашение.\nВы можете начать настройку через команду /setup.
 messages.destruction=Это сообщение самоудалится через %MIN% минут(ы).
 selfCleanup.leave=Вы не настроили меня.\nЯ остаюсь только на активных серверах, в которых есть хотя бы один канал для репутации. Если хотите использовать меня вновь - можете в любой момент пригласить меня по этой ссылке [link](%INVITE%).
-selfCleanup.leave.title=Пока! o/
+selfCleanup.leave.title=Текстовые команды больше не поддерживаются. Вместо них используй команды со слешем.
```
Comment (Taucher2003) [inline src/main/resources/locale_ru_RU.properties:246]: This is the translation from `error.textCommand` (Suggesting revert)
```suggestion
selfCleanup.leave.title=Пока! o/
```

**Review by Taucher2003 — CHANGES_REQUESTED** [review]

File: src/main/java/de/chojo/repbot/commands/TopMonth.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import static de.chojo.repbot.commands.Top.buildTop;
+
+public class TopMonth extends SimpleCommand {
+    private static final int TOP_PAGE_SIZE = 10;
+    private final ReputationData reputationData;
+    private final Localizer loc;
+
+    public TopMonth(DataSource dataSource, Localizer localizer) {
+        super("topmonth",
+                new String[]{"reptop"},
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/TopMonth.java:24]: Does this result in a conflict?

File: src/main/java/de/chojo/repbot/commands/TopWeek.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import static de.chojo.repbot.commands.Top.buildTop;
+
+public class TopWeek extends SimpleCommand {
+    private static final int TOP_PAGE_SIZE = 10;
+    private final ReputationData reputationData;
+    private final Localizer loc;
+
+    public TopWeek(DataSource dataSource, Localizer localizer) {
+        super("topweek",
+                new String[]{"reptop"},
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/TopWeek.java:28]: Same here, does this result in a conflict?

File: src/main/java/de/chojo/repbot/commands/TopMonth.java (hunk 3 above)
Comment (Taucher2003) [reply src/main/java/de/chojo/repbot/commands/TopMonth.java:24]: This is only used for legacy text commands
```suggestion
                null,
```

File: src/main/java/de/chojo/repbot/commands/TopWeek.java (hunk 4 above)
Comment (Taucher2003) [reply src/main/java/de/chojo/repbot/commands/TopWeek.java:28]: This is only used for legacy text commands
```suggestion
                null,
```

File: build.gradle.kts [hunk 5]
```diff
@@ -15,11 +15,11 @@ repositories {
 
 dependencies {
     // discord
-    implementation("net.dv8tion", "JDA", "4.3.0_339") {
+    implementation("net.dv8tion", "JDA", "5.0.0-alpha.4") {
         exclude(module = "opus-java")
     }
 
-    implementation("de.chojo", "cjda-util", "1.5.5")
+    implementation("de.chojo", "cjda-util", "2.0.0a-DEV")
```
Comment (Taucher2003) [inline build.gradle.kts:22]:
```suggestion
    implementation("de.chojo", "cjda-util", "2.0.1")
```

**Review by Taucher2003 — APPROVED** [review]

**Pull request merged**